### PR TITLE
Adds fancy log messages for characters high on weed

### PR DESCRIPTION
**Pull request by gkarfakis19**
#### Summary
Interface "High characters now see fancy smiley faces and exclamations in their message log."

#### Purpose of change
Based on Erk's suggestion in #53152.
This will make being high more fun, introducing harmless (hopefully not too annoying) smiley faces and exclamations in the log.

#### Describe the solution

At most, you will see 3 smiley faces per message, and that is if you are *very* high. Most messages will have 0 to 1 smiley faces. The logic behind is a bit complex: I am only inserting the smiley faces between words, and they can only go after words that are long enough. The intent is to stop a smiley being added before words like "a" or "the", as they make it harder to read the log. Smiley faces can also go at the end of any chat message.

Furthermore, past a certain weed threshold, long messages may end in an exclamation, such as "Woah.", "Dude." (generic messages), "Wicked!" (good messages/critical hits) or "Bummer." (bad messages).
This is also inspired by Erk's suggestions.

I would appreciate feedback on whether this happens too often or not. I can certainly tone down the frequency of both smileys and explanations, but this seems reasonable to me and is a cool flavor effect.

#### Describe alternatives you've considered
Not doing this.

#### Testing
image
(image is old, I have made the smiley handling smarter so that their placement is more consistent).

### CODE REVIEW

**Comment by KeremBabaG** [comment]
peak cdda

**Comment by fairyarmadillo** [comment]
If we can get the message log to show up in ridiculous colors for people who are very high on acid or mushrooms, I would consider that a Christmas present to the entire community.

**Comment by gkarfakis19** [comment]
What's happening with these test failures? It seems that the build environment isn't even initialized properly.

**Review by ehughsbaird — COMMENTED** [review]
It would make the code a lot cleaner to move this to a separate function. It would mean that the function that's about adding a message is not mostly concerned with adding smiley faces for an odd one-off case.

This should also be done with an effect flag, not a specific id check.

**Comment by gkarfakis19** [comment]
> It would make the code a lot cleaner to move this to a separate function. It would mean that the function that's about adding a message is not mostly concerned with adding smiley faces for an odd one-off case.
> 
> This should also be done with an effect flag, not a specific id check.

I will move this to a separate function. I don't think we can do an effect flag, as I am directly looking at the duration of the specific effect to determine "intensity".

File: src/messages.h [hunk 1]
```diff
@@ -216,4 +216,6 @@ inline void add_msg_debug_if_player_sees( const Creature &target, debugmode::deb
     }
 }
 
+void modify_msg_with_exclamations( std::string &msg, game_message_type const type );
```
Comment (kevingranade) [inline src/messages.h:219]:
```suggestion
void modify_msg_with_exclamations( std::string &msg, game_message_type type );
```

File: src/messages.cpp [hunk 2]
```diff
@@ -239,6 +243,112 @@ class messages_impl
             }
         }
 
+        void modify_msg_with_exclamations( std::string &msg, game_message_type const type ) {
+            // add smileys to messages if the player is high on weed
+            Character *p_char = &get_player_character();
+            if( p_char->has_effect( effect_weed_high ) ) {
```
Comment (ehughsbaird) [inline src/messages.cpp:249]: Code style nag: you can drop a level of indentation (and make the code more readable) if you negate this condition and turn it into an early return.

Feel free to ignore.

**Pull request merged**